contracts/CryptoToFiat.sol

```diff
@@ -0,0 +1,143 @@
+    mapping(address => uint256) public lastTxnBlocks;
+
+    // To store the total conversion amount
+    mapping(address => uint256) public totalConvertedAmt;
+
+    //Tokens which have been deposit into the contract
+    mapping (address => uint256) public balances;
+    
+    // Address of token contract
+    ERC20 public token;
+
+    // Events
+    event PriceUpdated(string price);
+    event NewOwner(address owner);
+    event ConfigurationUpdate(uint256 minBalance, uint256 conversionUpperLimit, uint256 txnLimitInBlocks);
+    event DepositFunds(address indexed sender, uint256 amount);
+    event WithdrawFunds(address indexed sender, uint256 amount);
+    event TransferFunds(address indexed sender, address indexed receiver, uint256 amount);
+    event ConvertFunds(address indexed sender, address indexed owner, uint256 amount);
+    
+    // Modifiers
+    modifier onlyOwner() {
+        require(
+            msg.sender == owner,
+            "Only owner can call this function."
+        );
+        _;
+    }
+
+    constructor (address _token)
+    public
+    {
+        token = ERC20(_token);
+        owner = msg.sender;
+    }
+  
+    function deposit(uint256 value) 
+    public
+    returns(bool) 
+    {
+        require(token.transferFrom(msg.sender, this, value), "Unable to transfer token to the contract."); 
+        balances[msg.sender] = balances[msg.sender].add(value);
+        emit DepositFunds(msg.sender, value);
+        return true;
+    }
+    
+    function withdraw(uint256 value)
+    public
+    returns(bool)
+    {
+        require(balances[msg.sender] >= value, "Insufficient balance in the contract.");
+        require(token.transfer(msg.sender, value), "Unable to transfer token from the contract.");
+        balances[msg.sender] = balances[msg.sender].sub(value);
+        emit WithdrawFunds(msg.sender, value);
+        return true;
+    }
+    
+    function transfer(address receiver, uint256 value)
+    public
+    returns(bool)
+    {
+        require(balances[msg.sender] >= value, "Insufficient balance in the contract");
+        balances[msg.sender] = balances[msg.sender].sub(value);
+        balances[receiver] = balances[receiver].add(value);
+
+        emit TransferFunds(msg.sender, receiver, value);
+        return true;
+    }
+
+    function initiateConversion(uint256 value, uint256 totalClaim, uint8 v, bytes32 r, bytes32 s)
+    public
+    returns(bool)
+    {
+        require(balances[msg.sender] >= value, "Insufficient balance in the contract");
+        require(balances[msg.sender] >= value.add(minBalance), "Minimum balance to be maintained in the contract");
+        require(value <= conversionUpperLimit, "Exceeding the conversion limit");
+        require(lastTxnBlocks[msg.sender] <= block.number.sub(txnLimitInBlocks), "Exceeding the number of transactions in given time");
+        require(totalConvertedAmt[msg.sender].add(value) <= totalClaim, "Exceeding the claims made so far");
```

Comment:
From the message below I can deduce that `value` is the amount of tokens converted in the particular transaction and `totalClaim` is the total amount of tokens converted by sender. But then two questions appears:
1. Why do we need both `value` and `totalClaim` as parameters when one of them can be calculated from another?
2. This assert seems incorrect as following should be true: `totalConvertedAmt[msg.sender].add(value) == totalClaim`.

Reply:
@vsbogd totalClaim - Is  the total amount claimed so far in MPE contract. Value parameter is the amount the user want to convert.
totalConvertedAmt is where we track the amount converted using our contract. It should be always less than equal to totalClaim done in MPE contract.

Reply:
So each time when the service provider claims it receives the authorization from `authorizer` to convert less or equal amount of tokens in total to the fiat. Then the service provider can start converting tokens using the authorization but he can convert only while other conditions (like the transaction limit and the frequency) are met.

Is it correct explanation? If so I am not sure we need to stick to `totalClaim` semantics. Should it be just `totalAuthorizedAmount`?

Reply:
Yes you are right. Authorizer checks MPE and provides only the totalClaimable amount as one of the parameter.
totalConvertedAmt is to track how much the user converted so far. So whenever users converts the amount we keep adding it to this field.
We have require condiftion to make sure that the totalConvertedAmt cant be exceeding the totalClaim amount to prevent arbitrage trading.

Looks like the confusion is with the totalClaim parameter, I can change it to totalClaimable to make it more clear. Thanks!

Reply:
@ksridharbabuus, yes, may be change name and add documentation to explain it, thanks.

Reply:
Updated the contract and added comments to function as well.

Reply:
Updated the contract and added comments to function as well.